File: modules/SSHFleet_Py/src/security/credential.py
# -*- coding: utf-8 -*-
# SSHFleet 凭据深模块
# 职责：密码/口令类凭据「读盘 → 格式分类 → 安全等级匹配 → 解码/解密」的单一实现。
# 消除 csv.py 两套与 upgrade.py 一套的重复错配矩阵（架构审查候选 #2），
# 并缓存解码结果，避免校验阶段与使用阶段双重读盘/取主密钥（等级3时每节点只读盘一次）。
# 决策背景见 docs/adr/0007-homemade-cipher-env-key.md

import base64
import os
from typing import List, Optional, Tuple

from src.security.cipher import CipherError, classify_credential, decrypt_password, is_probably_base64_text, looks_encrypted
from src.security.master_key import get_master_key_or_exit

# 解码结果缓存：(path, level) -> 明文。进程内单次运行，仅加速不改变语义；
# 凭据文件运行中不会变化，无需失效；明文本就存在于流程内存（ADR-0007 不防内存明文）。
_decode_cache: dict = {}


def _check_fmt_level(fmt: str, level: str) -> Optional[str]:
    """凭据内容格式 × 密码安全等级的错配诊断（全工具单一事实来源）。

    返回错误码或 None（匹配）。错误码语义：
      mismatch_encrypted / mismatch_base64 / bad_cipher / bad_base64
    规则（与历史 _check_credential_file 一致，修正 _read_credential 缺失的 base64×3 分支）：
      encrypted 仅等级3匹配；base64 仅等级2匹配（等级3要求加密格式，报 bad_cipher）；
      plain 仅等级1匹配。
    """
    if fmt == "encrypted" and level != "3":
        return "mismatch_encrypted"
    if fmt == "base64" and level == "1":
        return "mismatch_base64"
    if fmt == "base64" and level == "3":
        return "bad_cipher"
    if fmt == "plain" and level == "2":
        return "bad_base64"
    if fmt == "plain" and level == "3":
        return "bad_cipher"
    return None


def decode_credential(content: str, level: str) -> Tuple[Optional[str], Optional[str], Optional[str]]:
    """按密码安全等级把凭据内容还原为明文（不读盘）。

    返回 (明文, 错误码, 错误细节)，明文与错误互斥：
      level=1 原样返回；level=2 base64 解码；level=3 主密钥解密。
    """
    fmt = classify_credential(content)
    code = _check_fmt_level(fmt, level)
    if code:
        return None, code, None
    if level == "1":
        return content, None, None
    if level == "3":
        master_key = get_master_key_or_exit("read_credential")
        try:
            return decrypt_password(content, master_key), None, None
        except CipherError as e:
            return None, "bad_cipher", str(e)
    try:
        return base64.b64decode(content).decode("utf-8"), None, None
    except Exception:
        return None, "bad_base64", None


def read_credential(path: str, level: str, require_nonempty: bool = False) -> Tuple[Optional[str], List[Tuple[str, Optional[str]]]]:
    """密码/口令类凭据：读盘 → 判空 → 格式分类 → 等级匹配 → 解码/解密（带缓存）。

    返回 (明文, 错误码列表)，空列表=通过；错误码与历史 _check_credential_file 兼容：
      missing / read_error / empty / bad_base64 / bad_cipher / mismatch_encrypted / mismatch_base64
    成功时结果按 (path, level) 缓存，后续读取（如逐节点解析）不再读盘/取主密钥。
    """
    cache_key = (path, level)
    if cache_key in _decode_cache:
        return _decode_cache[cache_key], []

    if not os.path.exists(path):
        return None, [("missing", None)]
    try:
        with open(path, "r", encoding="utf-8") as f:
            content = f.read().strip()
    except Exception as e:
        return None, [("read_error", str(e))]
    if not content:
        return None, [("empty", None)]

    plain, code, detail = decode_credential(content, level)
    if code:
        return None, [(code, detail)]
    if require_nonempty and not plain:
        return None, [("empty_decoded", None)]

    _decode_cache[cache_key] = plain
    return plain, []


def read_credential_pem(path: str) -> Tuple[Optional[str], List[Tuple[str, Optional[str]]]]:
    """私钥 PEM 读取校验：读盘 → 判空 → -----BEGIN 前缀检查（不参与 fmt×level 矩阵）。

    返回 (内容, 错误码列表)；错误码：missing / read_error / empty / bad_pem
    """
    if not os.path.exists(path):
        return None, [("missing", None)]
    try:
        with open(path, "r", encoding="utf-8") as f:
            content = f.read().strip()
    except Exception as e:
        return None, [("read_error", str(e))]
    if not content:
        return None, [("empty", None)]
    if not content.startswith("-----BEGIN"):
        return None, [("bad_pem", None)]
    return content, []


# ---- 错误码 → 用户文案（单一事实来源，校验汇总 / 退出指引两条路径共用） ----
# 本表贴着错误码的产生地（上方读取函数）放置。收敛前 csv.py 内 if 链与字典两张文案表
# 已措辞漂移（mismatch_encrypted），且互相缺码：if 链缺 empty_decoded/bad_pem
# （静默落 bad_base64 文案），字典无兜底（未知码直接 KeyError）。

_CREDENTIAL_LABELS = {
    "missing": "不存在",
    "read_error": "无法读取",
    "empty": "内容为空",
    "bad_base64": "不是有效的Base64编码",
    "empty_decoded": "解码后内容为空",
    "bad_pem": "不是有效的PEM格式（缺少 -----BEGIN 头）",
    "bad_cipher": "不是有效的加密格式或主密钥不匹配（请先用 --convert-password 转换该文件）",
    "mismatch_encrypted": "文件是本工具等级3（加密）格式，与当前密码安全等级不匹配",
    "mismatch_base64": "文件是等级2（base64）格式，与当前密码安全等级不匹配",
}


def credential_error_label(code: str, path: str, detail: Optional[str] = None) -> str:
    """凭据错误码 → 汇总短文案（CSV 校验汇总路径用）。未知码兜底，不再 KeyError。"""
    label = _CREDENTIAL_LABELS.get(code, f"未知凭据错误({code})")
    if detail:
        return f"{label} → {path} ({detail})"
    return f"{label} → {path}"


def credential_error_detail(code: str, level: str, path: str, detail: Optional[str]) -> str:
    """凭据错误码 → 退出前完整指引文案（读取失败即退路径用）。未知码兜底，不再误标为 bad_base64。"""
    if code == "missing":
        return f"凭据文件不存在：{path}"
    if code == "read_error":
        return f"凭据文件无法读取：{path} ({detail})"
    if code == "empty":
        return f"凭据文件内容为空：{path}"
    if code == "mismatch_encrypted":
        return (
            f"凭据文件是本工具等级3（加密）格式，与当前密码安全等级 {level}（1=明文 2=base64 3=加密）不匹配：{path}\n"
            f"请将配置 account.password_security 改为 3，或先用 --convert-password 处理该文件"
        )
    if code == "mismatch_base64":
        return (
            f"凭据文件是等级2（base64）格式，与当前密码安全等级 1（明文）不匹配：{path}\n"
            f"请先将该文件内容还原为明文，或将配置改为 2"
        )
    if code == "bad_cipher":
        if detail:  # 解密失败（密钥不匹配/文件损坏）
            return (
                f"凭据文件解密失败（文件可能尚未用 --convert-password 转换，或主密钥不匹配）：{path}\n{detail}"
            )
        return f"凭据文件不是等级{level}（加密）格式，请先 --convert-password 转换：{path}"
    if code == "bad_base64":
        return f"凭据文件不是等级{level}（base64）格式（内容疑似明文），请先 --convert-password 转换：{path}"
    if code == "empty_decoded":
        return f"凭据文件内容解码后为空，请检查文件是否填入了有效密码：{path}"
    if code == "bad_pem":
        return f"凭据文件不是有效的PEM格式（缺少 -----BEGIN 头）：{path}"
    return f"凭据文件校验失败（{code}）：{path}"


def read_cred_file_content(path: str) -> str:
    """凭据文件内容读取（--convert-password 转换场景）：读盘 + 判空 + 文本/单行防御。

    与原 upgrade._read_cred_content 行为一致，供转换流程复用；
    错误直接 print_error_information_and_exit 退出（转换流程约定）。

    Raises:
        SystemExit: 读盘失败 / 空文件 / 含二进制数据 / 多行非 base64 或加密格式
    """
    from src.common.error_handler import print_error_information_and_exit

    try:
        with open(path, "r", encoding="utf-8") as f:
            content = f.read().strip()
    except Exception as e:
        print_error_information_and_exit(
            "handle_convert_password", f"读取凭据文件失败：{path}\n原因：{e}"
        )
    if not content:
        print_error_information_and_exit(
            "handle_convert_password", f"凭据文件是空的，请先填入密码再转换：{path}"
        )
    # 防御：含 NUL 字节视为二进制文件，明确提示而非当作文本处理
    if "\x00" in content:
        print_error_information_and_exit(
            "handle_convert_password",
            f"凭据文件含二进制数据，不是文本格式，无法转换：{path}",
        )
    # 防御：明文凭据应为单行；多行 base64 / 加密格式（去空白后合法存储形态）放行
    if "\n" in content or "\r" in content:
        if not (is_probably_base64_text(content) or looks_encrypted(content)):
            print_error_information_and_exit(
                "handle_convert_password",
                f"凭据文件有多行内容，但密码应为单行，请检查是否误粘贴：{path}",
            )
    return content

File: modules/SSHFleet_Go/internal/credential/credential.go
// Package credential 承载主干第 4 步（工具模式分流）与凭据读写：
// 三等级凭据、主密钥（SSHFLEET_KEY）、--convert-password 与 0x02 AEAD 加密
// （spec D15/D30/D31；2026-09-14 裁定移除 0x01 旧密文兼容，见 D14 修订注）。
package credential

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"unicode/utf8"
)

// CredCode 凭据错误码（自定义字符串类型 + 常量，spec 实现途径 8：拼错编译不过）。
type CredCode string

const (
	CodeMissing           CredCode = "missing"
	CodeReadError         CredCode = "read_error"
	CodeEmpty             CredCode = "empty"
	CodeBadBase64         CredCode = "bad_base64"
	CodeEmptyDecoded      CredCode = "empty_decoded"
	CodeBadPEM            CredCode = "bad_pem"
	CodeBadCipher         CredCode = "bad_cipher"
	CodeLegacyCipher      CredCode = "legacy_cipher"
	CodeMismatchEncrypted CredCode = "mismatch_encrypted"
	CodeMismatchBase64    CredCode = "mismatch_base64"
)

// CredError 错误码 + 细节。
type CredError struct {
	Code   CredCode
	Detail string
}

// checkFmtLevel 凭据内容格式 × 密码安全等级的错配诊断（全工具单一事实来源）。
// encrypted 仅等级3匹配；base64 仅等级2匹配（等级3要求加密格式，报 bad_cipher）；plain 仅等级1匹配。
func checkFmtLevel(fmtStr string, level int) (CredCode, bool) {
	switch {
	case fmtStr == "encrypted" && level != 3:
		return CodeMismatchEncrypted, true
	case fmtStr == "base64" && level == 1:
		return CodeMismatchBase64, true
	case fmtStr == "base64" && level == 3:
		return CodeBadCipher, true
	case fmtStr == "plain" && level == 2:
		return CodeBadBase64, true
	case fmtStr == "plain" && level == 3:
		return CodeBadCipher, true
	}
	return "", false
}

// decodeCredential 按密码安全等级把凭据内容还原为明文（不读盘）。
// 返回 (明文, 错误码, 错误细节, 致命错误)。明文与错误互斥；
// 等级 3 且内容确为加密格式时才拉取主密钥（懒获取，对位旧 get_master_key_or_exit 时机）。
// 旧 0x01 密文不再解密：识别为 legacy_cipher，给「不再支持」的明确报错（2026-09-14 裁定）。
func decodeCredential(content string, level int) (string, CredCode, string, error) {
	fmtStr := ContentFormat(content)
	if code, bad := checkFmtLevel(fmtStr, level); bad {
		return "", code, "", nil
	}
	switch level {
	case 1:
		return content, "", "", nil
	case 3:
		if looksEncryptedV1(content) {
			return "", CodeLegacyCipher, "", nil
		}
		masterKey, fatalErr := GetMasterKey()
		if fatalErr != nil {
			return "", "", "", fatalErr
		}
		plain, err := DecryptV2(content, masterKey)
		if err != nil {
			return "", CodeBadCipher, err.Error(), nil
		}
		return plain, "", "", nil
	default: // level 2
		raw, err := decodeStrictB64(content)
		if err != nil {
			return "", CodeBadBase64, "", nil
		}
		if !utf8.Valid(raw) {
			return "", CodeBadBase64, "", nil
		}
		return string(raw), "", "", nil
	}
}

// ReadCredential 密码/口令类凭据的一条龙读取：读盘 → 判空 → 格式分类 → 等级匹配 → 解码/解密 → 出文案。
// requireNonempty 对位密码类校验（口令类不判空）。旧版 (path, level) 解码缓存不移植：
// 解码发生在预检、结果直接进节点数据（spec 实现层差异）。
// 返回 (明文, 问题文案列表, 致命错误)；列表空 = 通过。
// 文案每条已含路径，调用方只需拼上自己的前缀（行号 / IP / 列名），不必了解错误码。
func ReadCredential(path string, level int, requireNonempty bool) (string, []string, error) {
	plain, credErrs, fatalErr := readCredentialCore(path, level, requireNonempty)
	if fatalErr != nil {
		return "", nil, fatalErr
	}
	return plain, credProblems(credErrs, path), nil
}

// readCredentialCore 结构化读取核心（包内 seam）：错误保持错误码形态，文案统一由 ReadCredential 组装。
func readCredentialCore(path string, level int, requireNonempty bool) (string, []CredError, error) {
	if _, err := os.Stat(path); err != nil {
		return "", []CredError{{CodeMissing, ""}}, nil
	}
	content, err := os.ReadFile(path)
	if err != nil {
		return "", []CredError{{CodeReadError, err.Error()}}, nil
	}
	text := strings.TrimSpace(string(content))
	if text == "" {
		return "", []CredError{{CodeEmpty, ""}}, nil
	}
	plain, code, detail, fatalErr := decodeCredential(text, level)
	if fatalErr != nil {
		return "", nil, fatalErr
	}
	if code != "" {
		return "", []CredError{{code, detail}}, nil
	}
	if requireNonempty && plain == "" {
		return "", []CredError{{CodeEmptyDecoded, ""}}, nil
	}
	return plain, nil, nil
}

// ReadCredentialPEM 私钥 PEM 读取校验：读盘 → 判空 → -----BEGIN 前缀检查（不参与 fmt×level 矩阵）。
// 返回 (内容, 问题文案列表)，语义同 ReadCredential。
func ReadCredentialPEM(path string) (string, []string) {
	content, credErrs := readCredentialPEMCore(path)
	return content, credProblems(credErrs, path)
}

// readCredentialPEMCore 私钥 PEM 结构化读取核心（包内 seam）。
func readCredentialPEMCore(path string) (string, []CredError) {
	if _, err := os.Stat(path); err != nil {
		return "", []CredError{{CodeMissing, ""}}
	}
	content, err := os.ReadFile(path)
	if err != nil {
		return "", []CredError{{CodeReadError, err.Error()}}
	}
	text := strings.TrimSpace(string(content))
	if text == "" {
		return "", []CredError{{CodeEmpty, ""}}
	}
	if !strings.HasPrefix(text, "-----BEGIN") {
		return "", []CredError{{CodeBadPEM, ""}}
	}
	return text, nil
}

// ReadPEMRaw 私钥 PEM 原样读取（对位 _read_credential(decode_base64=False)）。
// 调用方需保证文件已通过 ReadCredentialPEM /  -k 存在性校验。
func ReadPEMRaw(path string) (string, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return "", err
	}
	return strings.TrimSpace(string(content)), nil
}

// ---- 错误码 → 用户文案（单一事实来源，校验汇总 / 退出指引两条路径共用） ----

// credProblems 错误码列表 → 调用方可直接拼接的短文案列表（每条已含路径），空 = 通过。
func credProblems(credErrs []CredError, path string) []string {
	if len(credErrs) == 0 {
		return nil
	}
	problems := make([]string, 0, len(credErrs))
	for _, ce := range credErrs {
		problems = append(problems, CredErrorLabel(ce.Code, path, ce.Detail))
	}
	return problems
}

// CredErrorLabel 凭据错误码 → 汇总短文案（CSV 校验汇总路径用）。
func CredErrorLabel(code CredCode, path string, detail string) string {
	var label string
	switch code {
	case CodeMissing:
		label = "不存在"
	case CodeReadError:
		label = "无法读取"
	case CodeEmpty:
		label = "内容为空"
	case CodeBadBase64:
		label = "不是有效的Base64编码"
	case CodeEmptyDecoded:
		label = "解码后内容为空"
	case CodeBadPEM:
		label = "不是有效的PEM格式（缺少 -----BEGIN 头）"
	case CodeBadCipher:
		label = "不是有效的加密格式或主密钥不匹配（请先用 --convert-password 转换该文件）"
	case CodeLegacyCipher:
		label = "是旧版 0x01 加密格式（5.0.0 起不再支持，请重新录入密码后按当前等级重新转换）"
	case CodeMismatchEncrypted:
		label = "文件是本工具等级3（加密）格式，与当前密码安全等级不匹配"
	case CodeMismatchBase64:
		label = "文件是等级2（base64）格式，与当前密码安全等级不匹配"
	default:
		label = fmt.Sprintf("未知凭据错误(%s)", code)
	}
	if detail != "" {
		return fmt.Sprintf("%s → %s (%s)", label, path, detail)
	}
	return fmt.Sprintf("%s → %s", label, path)
}

// ---- D31 路径解析单点：写在文件里的凭据相对路径拼 secret_dir ----

// ErrRelativeNoSecretDir 相对路径但 secret_dir 未配置（调用方按各自场景组织文案）。
var ErrRelativeNoSecretDir = errors.New("相对路径但 secret_dir 未配置")

// ResolveSecretPath 凭据路径梯子（全工具单一事实来源，spec D31 统一解析顺序）：
// 去首尾空白 → ~ 展开 → 绝对则原样 → 相对则拼 secret_dir。
// 存在性校验由调用方负责（CSV 场景经 ReadCredential 的 missing，转换场景显式 Stat）。
func ResolveSecretPath(raw, secretDir string) (string, error) {
	p := strings.TrimSpace(raw)
	p = expandHomeTilde(p)
	if filepath.IsAbs(p) {
		return p, nil
	}
	if secretDir == "" {
		return "", ErrRelativeNoSecretDir
	}
	return filepath.Join(secretDir, p), nil
}

// expandHomeTilde 把前导 ~ 展开为用户主目录。
func expandHomeTilde(p string) string {
	home, err := os.UserHomeDir()
	if err != nil {
		return p
	}
	if p == "~" {
		return home
	}
	if strings.HasPrefix(p, "~/") || strings.HasPrefix(p, `~\`) {
		return filepath.Join(home, p[2:])
	}
	return p
}

// ReadCredFileContent 凭据文件内容读取（--convert-password 转换场景）：
// 读盘 + 判空 + 文本/单行防御（对位旧 read_cred_file_content）。
func ReadCredFileContent(path string) (string, error) {
	content, err := os.ReadFile(path)
	if err != nil {
		return "", fmt.Errorf("读取凭据文件失败：%s\n原因：%v", path, err)
	}
	text := strings.TrimSpace(string(content))
	if text == "" {
		return "", fmt.Errorf("凭据文件是空的，请先填入密码再转换：%s", path)
	}
	if strings.ContainsRune(text, 0) {
		return "", fmt.Errorf("凭据文件含二进制数据，不是文本格式，无法转换：%s", path)
	}
	// 防御：明文凭据应为单行；多行 base64 / 加密格式（去空白后合法存储形态）放行
	if strings.ContainsAny(text, "\n\r") {
		if !(isProbablyBase64Text(text) || looksEncryptedV1(text) || looksEncryptedV2(text)) {
			return "", fmt.Errorf("凭据文件有多行内容，但密码应为单行，请检查是否误粘贴：%s", path)
		}
	}
	return text, nil
}
